aten/src/ATen/native/TensorTransformations.cpp

```diff
@@ -9,51 +10,66 @@
 namespace at {
 namespace native {
 
-Tensor flip_cpu(const Tensor& self, IntList dims) {
-  const int64_t total_dims = self.dim(), flip_dims_size = dims.size();
-  flip_check_errors(total_dims, flip_dims_size, dims);
-
-  auto flip_dims_v = dims.vec();
-  wrap_all_dims(flip_dims_v, total_dims);
-  std::sort(flip_dims_v.begin(), flip_dims_v.end());
-  auto final_indices = std::vector<at::Tensor>(total_dims);
-
-  auto indices = std::vector<at::Tensor>(flip_dims_size);
-  for (int64_t i = 0; i < flip_dims_size; i++) {
-    indices[i] = at::arange(self.size(flip_dims_v[i]) - 1, -1, -1, self.type().toScalarType(at::kLong));
-    // creates a meshgrid
-    auto temp = std::vector<int64_t>(flip_dims_size, 1);
-    temp[i] = indices[i].size(0);
-    indices[i] = indices[i].view(IntList(temp));
-    final_indices[flip_dims_v[i]] = indices[i];
-  }
-
-  // check if distance between two flip dims >= 2, where permute of output tensor is needed,
-  // because the advanced indexing puts all non-consecutive indices in the beginning of the tensor
-  bool to_permute = false;
-  int64_t first = flip_dims_v[0], second = flip_dims_v[0];
-  for (int64_t i = 1; i < flip_dims_size; i++) {
-    second = flip_dims_v[i];
-    if (second - first >= 2) {
-      to_permute = true;
-      break;
+constexpr size_t dim_bitset_size = 64;
+
+template <typename scalar_t>
+void inline flip_cpu_kernel(
+  const int64_t total_dims,
+  const int64_t* stride_contiguous_d,
+  const std::bitset<dim_bitset_size>& flip_dims_b,
+  const Tensor& in_tensor,
+  Tensor& out_tensor
+){
+  int64_t i;
+  const int64_t numel = in_tensor.numel();
+  const scalar_t* in_tensor_d = in_tensor.data<scalar_t>();
+  scalar_t* out_tensor_d = out_tensor.data<scalar_t>();
+
+  #pragma omp parallel for private(i) if (numel > 1000)
+  for (i = 0; i < numel; i++) {
+    int64_t cur_indices = i;
+    int64_t rem = 0;
+    int64_t dst_offset = 0;
+
+    for (int64_t d = 0; d < total_dims; d++) {
+      int64_t temp = cur_indices;
+      cur_indices = cur_indices / stride_contiguous_d[d];
+      rem = temp - cur_indices * stride_contiguous_d[d];
+      if (flip_dims_b[d]) cur_indices = in_tensor.size(d) - 1 - cur_indices;
```

Comment:
I would suggest rewrite this look content a bit:
1. don't change `cur_indices` basing on the flag, instead, do `dst_offset += flag ? sizes[d] - 1 - cur_indices : cur_indices`
2. get a vector of `in_tensor.sizes()` outside and use that.

After that, if would be good to do another benchmark.

Reply:
Thanks a lot! I updated the code according your suggestions and now the runtime gets very closed to the best perf as in V1 !
```
>>> a = torch.randn(1000, 1000)
>>> %timeit -r 100 a.flip(0, 1)
3.04 ms ± 912 µs per loop (mean ± std. dev. of 100 runs, 100 loops each)
```